### PR TITLE
Record more artifact sizes during self-profiling.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR adds artifact size recording for

- "linked artifacts" (executables, RLIBs, dylibs, static libs)
- object files
- dwo files
- assembly files
- crate metadata
- LLVM bitcode files
- LLVM IR files
- codegen unit size estimates

Currently the identifiers emitted for these are hard-coded as string literals. Is it worth adding constants to https://github.com/rust-lang/measureme/blob/master/measureme/src/rustc.rs instead? We don't do that for query names and the like -- but artifact kinds might be more stable than query names.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
@rylev, did I forget any artifact kind we talked about?

**Comment by rylev** [comment]
@michaelwoerister the only one that comes to mind is final binary size. Is there a convenient place for recording that after the linker has done its thing?

**Review by davidtwco — APPROVED** [review]
> Currently the identifiers emitted for these are hard-coded as string literals. Is it worth adding constants to https://github.com/rust-lang/measureme/blob/master/measureme/src/rustc.rs instead? We don't do that for query names and the like -- but artifact kinds might be more stable than query names.

I don't think it would hurt, if you want to do this, feel free, don't think it's strictly necessary, within the compiler each of these constants would probably only be used once? r=me regardless.

**Comment by michaelwoerister** [comment]
> I don't think it would hurt, if you want to do this, feel free, don't think it's strictly necessary, within the compiler each of these constants would probably only be used once? 

Yes they would only be used once in the compiler. However, it would be useful for keeping the compiler and measureme-based tooling in sync. I opened https://github.com/rust-lang/measureme/issues/191 so we don't forget. But it's not worth doing a new measureme release for that, I'd say.

Thanks for the reviews, @davidtwco & @rylev!

@bors r=davidtwco

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 1]
```diff
@@ -76,6 +78,19 @@ pub fn write_output_file(
...
+        if result == llvm::LLVMRustResult::Success {
+            let artifact_kind = match file_type {
+                llvm::FileType::ObjectFile => "object_file",
+                llvm::FileType::AssemblyFile => "assembly_file",
+            };
+            record_artifact_size(self_profiler_ref, artifact_kind, output);
+            if let Some(dwo_file) = dwo_output {
+                record_artifact_size(self_profiler_ref, "dwo_file", dwo_file);
+            }
+        }
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/src/back/write.rs:92]: Can you update https://github.com/rust-lang/rust/blob/60952bc3da442cf2ea11dcc9c80a461a659bbcd7/compiler/rustc_codegen_cranelift/src/driver/aot.rs#L39 too please?

Comment (bjorn3) [reply compiler/rustc_codegen_llvm/src/back/write.rs:92]: Done in bjorn3/rustc_codegen_cranelift@f3b2d37e7754458a14ac99cee970ede383eddc2f.

**Pull request merged**